### PR TITLE
Implement OffHeapMutableBytesStore for real-time var-length bytes store

### PR DESCRIPTION
**Pull request by Jackie-Jiang**
Comparing with the MutableOffHeapByteArrayStore:
1. Read/Write speed improvement
2. Buffer size is fixed and buffers are allocated on demand, so memory is always used effeciently
3. No need to estimate the buffer size beforehead, and performance is guaranteed (for MutableOffHeapByteArrayStore, if buffer size is under-estimated, pay huge penalty on allocating a huge buffer at runtime, and also read/write speed degration

Added benchmark to compare the read/write speed:
```
Benchmark                                                            (_maxValueLength)  Mode  Cnt     Score    Error  Units
BenchmarkOffHeapMutableBytesStore.mutableOffHeapByteArrayStoreRead                   8  avgt    5    29.875 ±  0.370  ms/op
BenchmarkOffHeapMutableBytesStore.mutableOffHeapByteArrayStoreRead                  16  avgt    5    32.344 ±  0.765  ms/op
BenchmarkOffHeapMutableBytesStore.mutableOffHeapByteArrayStoreRead                  32  avgt    5    35.027 ±  0.460  ms/op
BenchmarkOffHeapMutableBytesStore.mutableOffHeapByteArrayStoreRead                  64  avgt    5    40.811 ±  0.414  ms/op
BenchmarkOffHeapMutableBytesStore.mutableOffHeapByteArrayStoreRead                 128  avgt    5    51.767 ±  0.185  ms/op
BenchmarkOffHeapMutableBytesStore.mutableOffHeapByteArrayStoreRead                 256  avgt    5    77.065 ±  1.627  ms/op
BenchmarkOffHeapMutableBytesStore.mutableOffHeapByteArrayStoreRead                 512  avgt    5   127.167 ±  1.075  ms/op
BenchmarkOffHeapMutableBytesStore.mutableOffHeapByteArrayStoreRead                1024  avgt    5   283.145 ±  2.829  ms/op
BenchmarkOffHeapMutableBytesStore.mutableOffHeapByteArrayStoreWrite                  8  avgt    5    16.755 ±  0.374  ms/op
BenchmarkOffHeapMutableBytesStore.mutableOffHeapByteArrayStoreWrite                 16  avgt    5    25.579 ±  0.293  ms/op
BenchmarkOffHeapMutableBytesStore.mutableOffHeapByteArrayStoreWrite                 32  avgt    5    23.953 ±  0.507  ms/op
BenchmarkOffHeapMutableBytesStore.mutableOffHeapByteArrayStoreWrite                 64  avgt    5    51.442 ±  1.230  ms/op
BenchmarkOffHeapMutableBytesStore.mutableOffHeapByteArrayStoreWrite                128  avgt    5    64.334 ±  0.069  ms/op
BenchmarkOffHeapMutableBytesStore.mutableOffHeapByteArrayStoreWrite                256  avgt    5   260.415 ±  1.823  ms/op
BenchmarkOffHeapMutableBytesStore.mutableOffHeapByteArrayStoreWrite                512  avgt    5   615.528 ±  3.180  ms/op
BenchmarkOffHeapMutableBytesStore.mutableOffHeapByteArrayStoreWrite               1024  avgt    5  1168.779 ± 16.961  ms/op
BenchmarkOffHeapMutableBytesStore.offHeapMutableBytesStoreRead                       8  avgt    5    28.978 ±  0.535  ms/op
BenchmarkOffHeapMutableBytesStore.offHeapMutableBytesStoreRead                      16  avgt    5    33.420 ±  0.400  ms/op
BenchmarkOffHeapMutableBytesStore.offHeapMutableBytesStoreRead                      32  avgt    5    37.120 ±  0.272  ms/op
BenchmarkOffHeapMutableBytesStore.offHeapMutableBytesStoreRead                      64  avgt    5    39.552 ±  0.317  ms/op
BenchmarkOffHeapMutableBytesStore.offHeapMutableBytesStoreRead                     128  avgt    5    42.861 ±  0.588  ms/op
BenchmarkOffHeapMutableBytesStore.offHeapMutableBytesStoreRead                     256  avgt    5    47.025 ±  0.189  ms/op
BenchmarkOffHeapMutableBytesStore.offHeapMutableBytesStoreRead                     512  avgt    5    60.740 ±  0.677  ms/op
BenchmarkOffHeapMutableBytesStore.offHeapMutableBytesStoreRead                    1024  avgt    5    89.509 ±  1.838  ms/op
BenchmarkOffHeapMutableBytesStore.offHeapMutableBytesStoreWrite                      8  avgt    5    21.327 ±  0.509  ms/op
BenchmarkOffHeapMutableBytesStore.offHeapMutableBytesStoreWrite                     16  avgt    5    28.494 ±  0.304  ms/op
BenchmarkOffHeapMutableBytesStore.offHeapMutableBytesStoreWrite                     32  avgt    5    33.399 ±  0.533  ms/op
BenchmarkOffHeapMutableBytesStore.offHeapMutableBytesStoreWrite                     64  avgt    5    50.755 ±  1.700  ms/op
BenchmarkOffHeapMutableBytesStore.offHeapMutableBytesStoreWrite                    128  avgt    5    75.702 ±  1.079  ms/op
BenchmarkOffHeapMutableBytesStore.offHeapMutableBytesStoreWrite                    256  avgt    5   129.677 ±  1.942  ms/op
BenchmarkOffHeapMutableBytesStore.offHeapMutableBytesStoreWrite                    512  avgt    5   222.697 ±  2.109  ms/op
BenchmarkOffHeapMutableBytesStore.offHeapMutableBytesStoreWrite                   1024  avgt    5   392.053 ±  3.200  ms/op
```

### CODE REVIEW

File: pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/dictionary/OffHeapMutableBytesStore.java [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+    }
+
+    int valueBufferIndex = (_valueStartOffset + valueLength - 1) >>> VALUE_BUFFER_SIFT_OFFSET;
+    PinotDataBuffer valueBuffer;
+    // If the current value buffer does not have enough space, allocate a new buffer to store the value
+    if ((_valueStartOffset - 1) >>> VALUE_BUFFER_SIFT_OFFSET != valueBufferIndex) {
+      valueBuffer = _memoryManager.allocate(VALUE_BUFFER_SIZE, _allocationContext);
+      _valueBuffers[valueBufferIndex] = valueBuffer;
+      _totalBufferSize += VALUE_BUFFER_SIZE;
+      _valueStartOffset = valueBufferIndex << VALUE_BUFFER_SIFT_OFFSET;
```
Comment (mcvsubbu) [inline pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/dictionary/OffHeapMutableBytesStore.java]: Doesn't this statement limit MAX_NUM_BUFFERS to (Integer.SIZE - VALUE_BUFFER_SHIFT_OFFSET)?

File: pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/dictionary/OffHeapMutableBytesStore.java [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+      valueLength = valueEndOffset - previousValueEndOffset;
+    }
+
+    byte[] value = new byte[valueLength];
+    _valueBuffers[valueBufferIndex].copyTo(startOffsetInValueBuffer, value);
+    return value;
+  }
+
+  public boolean equals(int index, byte[] value) {
+    int valueLength = value.length;
```
Comment (mcvsubbu) [inline pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/dictionary/OffHeapMutableBytesStore.java]: final. Also, it may be useful to rename this as inputValueLength (and 'value' as 'inputValue') fpor readability

File: pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/dictionary/OffHeapMutableBytesStore.java [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+    } else {
+      // Not the first value in the value buffer
+      if (valueEndOffset - previousValueEndOffset != valueLength) {
+        return false;
+      }
+      startOffsetInValueBuffer = previousValueEndOffset & VALUE_BUFFER_MASK;
+    }
+
+    byte[] valueAtIndex = new byte[valueLength];
+    _valueBuffers[valueBufferIndex].copyTo(startOffsetInValueBuffer, valueAtIndex);
```
Comment (mcvsubbu) [inline pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/dictionary/OffHeapMutableBytesStore.java]: Wouldnt it be faster if you just compared by fetching the bytes from the valueBuffer?

File: pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/dictionary/OffHeapMutableBytesStore.java [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+  private static final int VALUE_BUFFER_SIZE = 1 << VALUE_BUFFER_SIFT_OFFSET;
+  private static final int VALUE_BUFFER_MASK = 0xFFFFFFFF >>> (Integer.SIZE - VALUE_BUFFER_SIFT_OFFSET);
+
+  private final PinotDataBuffer[] _offsetBuffers = new PinotDataBuffer[MAX_NUM_BUFFERS];
+  private final PinotDataBuffer[] _valueBuffers = new PinotDataBuffer[MAX_NUM_BUFFERS];
+  private final PinotDataBufferMemoryManager _memoryManager;
+  private final String _allocationContext;
+
+  private int _numValues;
+  private int _valueStartOffset;
```
Comment (mcvsubbu) [inline pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/dictionary/OffHeapMutableBytesStore.java]: Please add a comment on how this value is computed. If I understand right, it is the GLOBAL start offset of each value assuming a linear array of (VALUE_ BUFFER_SIZE * numBuffers). If a value cannot be held in one buffer, then the _valueStartOffset is set to the start of the next buffer.

File: pinot-core/src/test/java/org/apache/pinot/core/realtime/impl/dictionary/OffHeapMutableBytesStoreTest.java [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+        int index = RANDOM.nextInt(NUM_VALUES);
+        assertTrue(offHeapMutableBytesStore.equals(index, _values[index]));
+        if (!Arrays.equals(_values[index], _values[0])) {
+          assertFalse(offHeapMutableBytesStore.equals(0, _values[index]));
+          assertFalse(offHeapMutableBytesStore.equals(index, _values[0]));
+        }
+      }
+    }
+  }
+}
```
Comment (mcvsubbu) [inline pinot-core/src/test/java/org/apache/pinot/core/realtime/impl/dictionary/OffHeapMutableBytesStoreTest.java]: Please add a test that adds values in one thread and reads values from (say) 4 threads
Please specific tests that test the boundary conditions (hitting buffer as well as offset limits, or hitting one of the limits) rather than relying on the randomness.

File: pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/dictionary/OffHeapMutableBytesStore.java [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+ * <p>There are two sets of buffers allocated for the store, one set to store the value end offsets and the other set to
+ * store the values. The new buffer is allocated when the previous buffer is filled up. The first entry of each offset
+ * buffer stores the end offset of the last value (or 0 for the first offset buffer) so that we can always get the
+ * previous and current value's end offset from the same buffer.
+ */
+public class OffHeapMutableBytesStore implements Closeable {
+  private static final byte[] EMPTY_BYTES = new byte[0];
+
+  // MAX_NUM_BUFFERS = 8_192
+  private static final int MAX_NUM_BUFFERS = 1 << 13;
```
[the remark this answers is not in the document]
Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/dictionary/OffHeapMutableBytesStore.java]: This number decides how many total values we can store. 8K buffers allow at most 67M values of total size up to 8GB.
Added more comments on these numbers

File: pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/dictionary/OffHeapMutableBytesStore.java [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+
+  // OFFSET_BUFFER_SIZE = 32_768 + 4
+  private static final int OFFSET_BUFFER_SHIFT_OFFSET = 13;
+  private static final int OFFSET_BUFFER_SIZE = ((1 << OFFSET_BUFFER_SHIFT_OFFSET) + 1) << 2;
+  private static final int OFFSET_BUFFER_MASK = 0xFFFFFFFF >>> (Integer.SIZE - OFFSET_BUFFER_SHIFT_OFFSET);
+
+  // VALUE_BUFFER_SIZE = 1_048_576
+  private static final int VALUE_BUFFER_SIFT_OFFSET = 20;
```
[the remark this answers is not in the document]
Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/dictionary/OffHeapMutableBytesStore.java]: Good catch

File: pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/dictionary/OffHeapMutableBytesStore.java (hunk 4 above)
Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/dictionary/OffHeapMutableBytesStore.java]: Added

File: pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/dictionary/OffHeapMutableBytesStore.java [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+      offsetBuffer = _offsetBuffers[offsetBufferIndex];
+    }
+
+    int valueLength = value.length;
+    if (valueLength == 0) {
+      offsetBuffer.putInt((offsetIndex + 1) << 2, _valueStartOffset);
+      return _numValues++;
+    }
+
+    int valueBufferIndex = (_valueStartOffset + valueLength - 1) >>> VALUE_BUFFER_SIFT_OFFSET;
```
[the remark this answers is not in the document]
Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/dictionary/OffHeapMutableBytesStore.java]: We don't allow storing one value cross multiple buffers, so this is the buffer index for the whole value. Added more comments in the class header.

File: pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/dictionary/OffHeapMutableBytesStore.java [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+    PinotDataBuffer valueBuffer;
```
[the remark this answers is not in the document]
Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/dictionary/OffHeapMutableBytesStore.java]: Preconditions at this level can hurt the performance. This bytes store can hold 8GB of data, so we should never exceed MAX_NUM_BUFFERS

File: pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/dictionary/OffHeapMutableBytesStore.java (hunk 1 above)
Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/dictionary/OffHeapMutableBytesStore.java]: No, these 2 are independent.
We use the shift operation to quickly get the startOffset because it is the first value in the current value buffer.

File: pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/dictionary/OffHeapMutableBytesStore.java (hunk 2 above)
Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/dictionary/OffHeapMutableBytesStore.java]: I'm not a fan of using final on local variables. Changed the valueLength to inputValueLength.

File: pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/dictionary/OffHeapMutableBytesStore.java (hunk 3 above)
Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/dictionary/OffHeapMutableBytesStore.java]: You mean byte-by-byte comparison? If we reach here, that means there is a pretty high chance that the two values are the same (we call this method when hash code matches, and we can only reach here if the length also matches), thus doing batch read should be preferable. I added one extra check to optimize the case of small reads to compare byte-by-byte.

Comment (mcvsubbu) [reply pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/dictionary/OffHeapMutableBytesStore.java]: But then there is extra garbage generated in the read.

Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/dictionary/OffHeapMutableBytesStore.java]: True, but that is short lived garbage, and the benchmark in #4362 already proved that batch read can give big performance gain for big values.

**Comment by Jackie-Jiang** [comment]
@mcvsubbu Addressed all the comments, please take another look

File: pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/dictionary/OffHeapMutableBytesStore.java (hunk 1 above)
Comment (mcvsubbu) [reply pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/dictionary/OffHeapMutableBytesStore.java]: valueBufferIndex should fit in (Integer.SIZE - VALUE_BUFFER_SHIFT_OFFSET - 1) bits. Therefore, MAX_NUM_BUFFERS should be at most (2 ** (Integer.SIZE - VALUE_BUFFER_SHIFT_OFFSET - 1)).

If valueBufferIndex does not fit in these bits, then we have a problem with overflow in the expression (valueBufferIndex << VALUE_BUFFER_SHIFT_OFFSET)

Right?

File: pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/dictionary/OffHeapMutableBytesStore.java [hunk 10]
```diff
@@ -0,0 +1,252 @@
...
+
+/**
+ * Off-heap variable length mutable bytes store.
+ * <p>The class is thread-safe for single writer and multiple readers.
+ * <p>There are two sets of buffers allocated for the store:
+ * <ul>
+ *   <li>
+ *     Offset buffer: stores the value end offsets. New offset buffer is allocated when the previous buffer is filled
+ *     up. The first entry of each offset buffer stores the end offset of the last value (or 0 for the first offset
+ *     buffer) so that we can always get the previous and current value's end offset from the same buffer.
```
Comment (mcvsubbu) [inline pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/dictionary/OffHeapMutableBytesStore.java]: Useful to define "offset" here. I believe you mean the global offset and not the offset within the buffer. I am not sure how to describe "global offset" but do the best you can. At least write that the offset we are talking about is NOT the offset within a valueBuffer.

File: pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/dictionary/OffHeapMutableBytesStore.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (mcvsubbu) [reply pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/dictionary/OffHeapMutableBytesStore.java]: How can it hold 8GB of data when we store global offsets in 32 bits? Maybe I am missing something

File: pinot-core/src/test/java/org/apache/pinot/core/realtime/impl/dictionary/OffHeapMutableBytesStoreTest.java [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+      futures.add(executorService.submit(() -> {
+        for (int i = 0; i < NUM_VALUES; i++) {
+          offHeapMutableBytesStore.add(_values[i]);
+        }
+      }));
+
+      // Reader threads
+      for (int i = 0; i < numReaders; i++) {
+        futures.add(executorService.submit(() -> {
+          for (int j = 0; j < NUM_READS; j++) {
```
Comment (mcvsubbu) [inline pinot-core/src/test/java/org/apache/pinot/core/realtime/impl/dictionary/OffHeapMutableBytesStoreTest.java]: Instead of stopping with NUM_READS, stop when the writer thread is done.

File: pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/dictionary/OffHeapMutableBytesStore.java [hunk 12]
```diff
@@ -0,0 +1,252 @@
...
+@ThreadSafe
+public class OffHeapMutableBytesStore implements Closeable {
+  private static final byte[] EMPTY_BYTES = new byte[0];
+
+  // Each offset buffer holds 8_192 end offsets and one extra end offset for the last value in the previous buffer (or 0
+  // for the first buffer)
+  // OFFSET_BUFFER_SIZE = (8_192 + 1) * 4 = 32_772
+  private static final int OFFSET_BUFFER_SHIFT_OFFSET = 13;
+  private static final int OFFSET_BUFFER_SIZE = ((1 << OFFSET_BUFFER_SHIFT_OFFSET) + 1) << 2;
+  private static final int OFFSET_BUFFER_MASK = 0xFFFFFFFF >>> (Integer.SIZE - OFFSET_BUFFER_SHIFT_OFFSET);
```
Comment (mcvsubbu) [inline pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/dictionary/OffHeapMutableBytesStore.java]: Is'nt  (1<<OFFSET_BUFFER_SHIFT_OFFSET)-1 easier to understand?

File: pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/dictionary/OffHeapMutableBytesStore.java (hunk 1 above)
Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/dictionary/OffHeapMutableBytesStore.java]: Oh I see what you mean. Very good question. Yes, we should not store more than 2GB data in this bytes store, or we might trigger the int overflow for the global offset. This limitation is listed in the comments

File: pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/dictionary/OffHeapMutableBytesStore.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/dictionary/OffHeapMutableBytesStore.java]: I realized that, so I changed the upper bound to 2GB :-P

File: pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/dictionary/OffHeapMutableBytesStore.java (hunk 1 above)
Comment (mcvsubbu) [reply pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/dictionary/OffHeapMutableBytesStore.java]: So, if we can add a pre-conditions check in the constructor that will be useful.  As written, it seems to imply that the static definitions can be set to whatever and things will work.

File: pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/dictionary/OffHeapMutableBytesStore.java (hunk 10 above)
Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/dictionary/OffHeapMutableBytesStore.java]: Added

File: pinot-core/src/test/java/org/apache/pinot/core/realtime/impl/dictionary/OffHeapMutableBytesStoreTest.java (hunk 11 above)
Comment (Jackie-Jiang) [reply pinot-core/src/test/java/org/apache/pinot/core/realtime/impl/dictionary/OffHeapMutableBytesStoreTest.java]: Done

File: pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/dictionary/OffHeapMutableBytesStore.java (hunk 1 above)
Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/dictionary/OffHeapMutableBytesStore.java]: The limitation comes from the int type for offset, but not these constants. Meaning it still works for any buffer size/number of buffers, but we can only hold 2GB data.

**Comment by Jackie-Jiang** [comment]
> It will be nice if you can add a single test case in which both valueBuffer and offsetBuffer limits are reached in the same add() call, instead of leaving it to happen at random. You may need to add getters to expose the limits of the buffers, but it is worth it, I believe.

This 2 buffers are handled independently (first process offsetBuffer, then valueBuffer), so I believe it is not necessary to test that explicitly. Also, with 100_000 writes per test, there should be pretty high chance to hit this scenario.
